File: LeetCode/Explore/TopInterview/Blind75/Tree/102_binary_tree_level_order_traversal.cpp
// Recursion

// Complexity Analysis
// Time complexity : \mathcal{O}(N)O(N) since each node is processed exactly once.
// Space complexity : \mathcal{O}(N)O(N) to keep the output structure which contains N node values.

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
  
private:
  vector<vector<int> > vp;
  void preOrder(TreeNode *root, int level, vector<vector<int> > &vp) {
    if (!root) { // base case
      return;
    }
    if (level == vp.size()) {
      vp.push_back({root->val});
    }
    else {
      vp[level].push_back(root->val);
    }
    if (root->left) {
      preOrder(root->left, level + 1, vp);
    }
    if (root->right) {
      preOrder(root->right, level + 1, vp);
    }
  }
  
public:
  vector<vector<int>> levelOrder(TreeNode* root) {
    if (!root) {
      return vp;
    }
    preOrder(root, 0, vp);
    return vp;
  }
};

File: LeetCode/Explore/TopInterview/Blind75/Tree/102_binary_tree_level_order_traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
  
    def preOrder(self, root: Optional[TreeNode], level: int, ans: List[List[int]]):
      if root == None:
        return
      if len(ans) == level:
        ans.append([root.val])
      else:
        ans[level].append(root.val)

      if root.left:
        self.preOrder(root.left, level + 1, ans)
      
      if root.right:
        self.preOrder(root.right, level + 1, ans)
  
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        ans = list()
        self.preOrder(root, 0, ans)
        return ans
